### PR TITLE
Added 'distance' weight to car profile

### PR DESCRIPTION
**Pull request by oxidase**
# Issue

For #3638 added usage of `forward_rate` and `backward_rate` in `car.lua` profile. Also fixes weight assembly :bug: in `assembleLeg`

## Tasklist
 - [x] review
 - [x] adjust for comments

### CODE REVIEW

**Comment by emiltin** [comment]
can we do this in the appropriate handlers, so that we don't duplicate code between profiles?

**Comment by oxidase** [comment]
@emiltin moved to `Handlers.handle_weights`

**Comment by emiltin** [comment]
nice. should weights only be set in forward/backward directions if the way is  routable in the corresponding directions?

File: profiles/lib/handlers.lua [hunk 1]
```diff
@@ -399,15 +399,22 @@ function Handlers.handle_oneway(way,result,data,profile)
   end
 end
 
+function Handlers.handle_weights(way,result,data,profile)
+  if properties.weight_name == 'distance' then
+    -- set weight rates to 1 for the distance weight, edge weights are distance / rate
+    result.forward_rate = 1
```
Comment (TheMarex) [inline profiles/lib/handlers.lua]: As @emiltin pointed out this needs a check for either `{foward,backward}_speed > 0` or `{forward,backward}_mode != mode.inaccessible`.

File: include/engine/guidance/assemble_leg.hpp [hunk 2]
```diff
@@ -204,7 +207,7 @@ inline RouteLeg assembleLeg(const datafacade::BaseDataFacade &facade,
 
     return RouteLeg{std::round(distance * 10.) / 10.,
                     duration / 10.,
-                    duration / facade.GetWeightMultiplier(),
+                    weight / facade.GetWeightMultiplier(),
```
Comment (TheMarex) [inline include/engine/guidance/assemble_leg.hpp]: Urg, good catch.

File: profiles/lib/handlers.lua (hunk 1 above)
Comment (oxidase) [inline profiles/lib/handlers.lua]: Good point!  But if `{forward,backward}_mode != mode.inaccessible` then assertion `parsed_way.duration > 0 || parsed_way.forward_speed > 0` in `ProcessWay` will fail if speed is 0,
so the condition "`{forward,backward}_mode != mode.inaccessible` and `{foward,backward}_speed > 0`" is asserted to be true and if `{forward,backward}_mode == mode.inaccessible`  the weight or rate is not used to compute the edge weight.

What i missed is an additional assertion in `ProcessWay` if `parsed_way.forward_travel_mode != TRAVEL_MODE_INACCESSIBLE` and `!fallback_to_duration` then must be asserted `parsed_way.weight > 0 || parsed_way.forward_rate > 0`. At the moment such edges will have default initialized weights that is 0 for `ValueByEdge` and `ValueByMeter`.

**Comment by oxidase** [comment]
@emiltin for `{forward,backward}_mode != mode.inaccessible` 0 speed values will result in an assertion [here](https://github.com/Project-OSRM/osrm-backend/blob/8f28d317cf14aa80a3fe3b732d3b85743e75803b/src/extractor/extractor_callbacks.cpp#L133) or [here](https://github.com/Project-OSRM/osrm-backend/blob/8f28d317cf14aa80a3fe3b732d3b85743e75803b/src/extractor/extractor_callbacks.cpp#L150)
Also using 0 or negative rate to mark inaccessible edges if mode is not mode.inaccessible will be incorrect, i  added assertions to prevent such situations.

File: profiles/lib/handlers.lua (hunk 1 above)
Comment (emiltin) [inline profiles/lib/handlers.lua]: but why are forward and backward rates still being set, without checking whether the way is accessible (mode != 0 and speed > 0)? if the way is not accessible, i don't think the rate should be set.

Comment (oxidase) [inline profiles/lib/handlers.lua]: agree, i added checks and also set `result.weight` explicitly to 0, that means "use rate values to compute edge weights"

**Comment by TheMarex** [comment]
Travis had a hiccup here, restarted the build. Otherwise this looks good to merge. 👍

**Pull request merged**